physicalplan/function/function.go

```diff
@@ -1,101 +1,279 @@
+		return ignoreEmpty(f.Points, sample(f.Labels, f.StepTime, f.Points, func(points []promql.Point) float64 {
 			return points[0].V
 		}))
 	},
-	"rate": func(labels labels.Labels, points []promql.Point, stepTime time.Time, selectRange time.Duration) promql.Sample {
-		return ignoreSingleValue(points, sample(labels, stepTime, points, func(points []promql.Point) float64 {
-			return extrapolatedRate(points, true, true, stepTime, selectRange)
+	"rate": func(f FunctionArgs) promql.Sample {
+		return ignoreSingleValue(f.Points, sample(f.Labels, f.StepTime, f.Points, func(points []promql.Point) float64 {
+			return extrapolatedRate(points, true, true, f.StepTime, f.SelectRange)
 		}))
 	},
-	"delta": func(labels labels.Labels, points []promql.Point, stepTime time.Time, selectRange time.Duration) promql.Sample {
-		return ignoreSingleValue(points, sample(labels, stepTime, points, func(points []promql.Point) float64 {
-			return extrapolatedRate(points, false, false, stepTime, selectRange)
+	"delta": func(f FunctionArgs) promql.Sample {
+		return ignoreSingleValue(f.Points, sample(f.Labels, f.StepTime, f.Points, func(points []promql.Point) float64 {
+			return extrapolatedRate(points, false, false, f.StepTime, f.SelectRange)
 		}))
 	},
-	"increase": func(labels labels.Labels, points []promql.Point, stepTime time.Time, selectRange time.Duration) promql.Sample {
-		return ignoreSingleValue(points, sample(labels, stepTime, points, func(points []promql.Point) float64 {
-			return extrapolatedRate(points, true, false, stepTime, selectRange)
+	"increase": func(f FunctionArgs) promql.Sample {
+		return ignoreSingleValue(f.Points, sample(f.Labels, f.StepTime, f.Points, func(points []promql.Point) float64 {
+			return extrapolatedRate(points, true, false, f.StepTime, f.SelectRange)
 		}))
 	},
-	"resets": func(labels labels.Labels, points []promql.Point, stepTime time.Time, selectRange time.Duration) promql.Sample {
-		return ignoreEmpty(points, sample(labels, stepTime, points, resets))
+	"resets": func(f FunctionArgs) promql.Sample {
+		return ignoreEmpty(f.Points, sample(f.Labels, f.StepTime, f.Points, resets))
+	},
+	"clamp": func(f FunctionArgs) promql.Sample {
+		return ignoreSingleScalar(f.Points, f.ScalarPoints, calculatedSample(f.Labels, f.StepTime, f.Points, f.ScalarPoints, clamp))
 	},
+	"clamp_min": func(f FunctionArgs) promql.Sample {
+		return ignoreEmptyScalar(f.Points, f.ScalarPoints, calculatedSample(f.Labels, f.StepTime, f.Points, f.ScalarPoints, clampMin))
+	},
+	"clamp_max": func(f FunctionArgs) promql.Sample {
+		return ignoreEmptyScalar(f.Points, f.ScalarPoints, calculatedSample(f.Labels, f.StepTime, f.Points, f.ScalarPoints, clampMax))
+	},
+}
+
+// functionSelector returns []model.StepVector after processing input with desired function.
+type functionSelector struct {
+	vectorPool    *model.VectorPool
+	funcExpr      *parser.Call
+	mint          int64
+	maxt          int64
+	step          int64
+	currentStep   int64
+	stepsBatch    int
```

Comment:
Do we need these parameters? Functions don't usually select directly from storage.

Reply:
Will remove!